### PR TITLE
Queue snapshots

### PR DESCRIPTION
**Pull request by marc-casperlabs**
This is a working version that allows dumping snapshots of the queue of a node by running

```console
$ kill -USR1 $NODE_PID
```

where `$NODE_PID` should actually the PID of the node that should dump its queue. It will create a `queue_dump.json` file in the working directoy of the node.

Example, dumping a node with a single event in it:

```console
$ jq < queue_dump.json 
{
  "NetworkIncoming": [],
  "Network": [],
  "Regular": [
    "AddressGossiper"
  ],
  "Api": []
}
```

(the `jq` is not necessary here, but allows filtering. I use it because it makes the output pretty :))

Currently only the name of the event is dumped, as all the inner fields are skipped. If more output it desired, `Serialize` should be derived on inner events, possibly with `skip_serialize` on problematic fields.

## Some jq examples

Counts of all event types:

```jq
jq 'map_values( map(keys[0] | {"type": ., weight: 1})| group_by(.type) | map ([.[0].type,(.|length)]) | map({(.[0]): .[1]}) )' queue_dump.json
```

Count each queue

```jq
jq 'map_values(map(keys[0]))' queue_dump.json
```

### CODE REVIEW

**Comment by goral09** [comment]
@marc-casperlabs How about we get this to compile and then squash all commits into one? If people are expected to keep it on the side (without merging to `master`) and use when necessary, it will be much easier to have it as a single commit that can be cherry-picked easily.

**Comment by marc-casperlabs** [comment]
Thinking about it, I think it's not harmful to merge this into master - the cost of having a few extra `Serialize` / `Deserialize` derives floating around. Actual runtime cost is is a single, never-taken branch on an atomic bool.

**Review by Fraser999 — COMMENTED** [review]
I'd like to see a short description of how to use this added to the node's README.

I'd prefer to feature-gate the functionality (really just the signal-handling code so that we don't query a bool on every event, not suggesting we feature-gate the bulk of the changes in this PR).  But that's not a blocker - leave as is if you prefer.

Other than these points, this looks good to me.

**Comment by marc-casperlabs** [comment]
> I'd prefer to feature-gate the functionality (really just the signal-handling code so that we don't query a bool on every event, not suggesting we feature-gate the bulk of the changes in this PR). But that's not a blocker - leave as is if you prefer.

I think that's absolutely unnecessary. We are talking about what I believe amounts to one or two CPU instructions, essentially never branching! That's not even remotely close to the threshold I would consider required to make feature gating worthwhile.

I can remove a log message somewhere to compensate :)

File: node/src/reactor/validator/tests/scenario.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+use casper_types::{account::Weight, PublicKey};
```
[the remark this answers is not in the document]
Comment (marc-casperlabs) [reply node/src/reactor/validator/tests/scenario.rs:1]: Oops. No, its not :)

**Comment by marc-casperlabs** [comment]
> @marc-casperlabs How about we get this to compile and then squash all commits into one? If people are expected to keep it on the side (without merging to master) and use when necessary, it will be much easier to have it as a single commit that can be cherry-picked easily.

This will no longer work, as we have more and more events and such implement `Serialize` to be dumpable. I'd argue that it is more trouble than it is worth at this point.

**Comment by goral09** [comment]
> > @marc-casperlabs How about we get this to compile and then squash all commits into one? If people are expected to keep it on the side (without merging to master) and use when necessary, it will be much easier to have it as a single commit that can be cherry-picked easily.
> 
> This will no longer work, as we have more and more events and such implement `Serialize` to be dumpable. I'd argue that it is more trouble than it is worth at this point.

Sure, that comment is over month old.

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**